docs/customruns.md

```diff
@@ -0,0 +1,413 @@
+controller running on the cluster that is responsible for watching and updating
+`CustomRun`s which reference their type. If no such controller is running, `CustomRun`s
+will have no `.status` value and no further action will be taken.
+
+## Configuring a `CustomRun`
+
+A `CustomRun` definition supports the following fields:
+
+- Required:
+  - [`apiVersion`][kubernetes-overview] - Specifies the API version, for example
+    `tekton.dev/v1beta1`.
+  - [`kind`][kubernetes-overview] - Identifies this resource object as a `CustomRun` object.
+  - [`metadata`][kubernetes-overview] - Specifies the metadata that uniquely identifies the
+    `CustomRun`, such as a `name`.
+  - [`spec`][kubernetes-overview] - Specifies the configuration for the `CustomRun`.
+    - [`customRef`](#1-specifying-the-target-custom-task-with-customref) - Specifies the type and
+      (optionally) name of the custom task type to execute.
+    - [`customSpec`](#2-specifying-the-target-custom-task-by-embedding-its-spec) - Embed the custom task resource spec
+      directly in a `CustomRun`.
+- Optional:
+  - [`timeout`](#specifying-timeout) - specifies the maximum duration of a single execution
+    of a `CustomRun`.
+  - [`retries`](#specifying-retries) - specifies the number of retures upon `CustomRun`failure.
+  - [`params`](#specifying-parameters) - Specifies the desired execution
+    parameters for the custom task.
+  - [`serviceAccountName`](#specifying-a-serviceaccount) - Specifies a `ServiceAccount`
+    object that provides custom credentials for executing the `CustomRun`.
+  - [`workspaces`](#specifying-workspaces) - Specifies the physical volumes to use for the
+    [`Workspaces`](workspaces.md) required by a custom task.
+
+[kubernetes-overview]:
+  https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields
+
+### Specifying the target Custom Task
+
+A custom task resource's `CustomSpec` may be directly embedded in the `CustomRun` or it may
+be referred to by a `CustomRef`. But, not both at the same time.
+
+1. [Specifying the target Custom Task with customRef](#1-specifying-the-target-custom-task-with-customref)
+   Referring a custom task (i.e. `CustomRef` ) promotes reuse of custom task definitions.
+
+2. [Specifying the target Custom Task by embedding its spec](#2-specifying-the-target-custom-task-by-embedding-its-spec)
+   Embedding a custom task (i.e. `CustomSpec` ) helps in avoiding name collisions with other users within the same namespace.
+   Additionally, in a pipeline with multiple embedded custom tasks, the details of entire pipeline can be fetched in a
+   single API request.
+
+#### Specifying the target Custom Task with customRef
+
+To specify the custom task type you want to execute in your `CustomRun`, use the
+`customRef` field as shown below:
+
+```yaml
+spec:
+  customRef:
+    apiVersion: example.dev/v1alpha1
+    kind: Example
+```
```

Comment:
Would it be possible to use more realistic examples? For some custom tasks, it might make sense to define the custom task once on the cluster and reference it multiple times, and for others, creating a new CustomRun might make more sense. Also, apiVersion should be v1beta1, and I suggest including `kind: CustomRun`.

Reply:
Thanks Lee! It would be great if you have any example in your mind XD. Aslo @ScrapCodes Would like to hear if you have any existing Custom Task in your mind that we usually use them by referring a predefined Custom Task instead of embedded one? vice versa!

Reply:
Thinking about this comment a bit more.. Giving good examples is a bit tricky. We don't want to write something that looks like you can actually use it, because we don't have any CustomRun controllers and we don't want users copy pasting the example and getting annoyed when it doesn't work. But at the same time I think the variable naming can "clue" ways we intend users to interact with this feature.

For example, here's a wait Task run (copied from experimental repo):
```
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  generateName: wait-run-
spec:
  ref:
    apiVersion: example.dev/v0
    kind: Wait
  params:
  - name: duration
    value: 10s
```
Maybe our example here could be:
```
apiVersion: tekton.dev/v1beta1
kind: CustomRun
spec:
  ref:
    apiVersion: example.dev/v0
    kind: MyCustomKind
  params:
  - name: duration
    value: 10s
```
Here, I'd imagine the user wouldn't want to specify a "name" in the custom task ref, because it's more useful to define duration on the run. (Duration is more of a runtime feature than an authoring time feature.)

The `pipeline-loop` custom task has some examples where it is useful to specify a name in the custom task ref (e.g. https://github.com/tektoncd/experimental/blob/main/pipeline-loops/examples/simplepipelineloop.yaml) and could be adapted similarly.

WDYT? This might be something we can iterate on in the future.

Reply:
Thanks Lee @lbernick, this is a very great discussion, I'm thinking about what the biggest benefit of defining unamed CustomRun would be:

I would say it's a benefit for Custom Task authors instead of Custom Task end users:

It enables Custom Task authors reconcile on `v1beta1.CustomRun` without defining a new CRD for the custom kind `MyCustomKind`.

Imagine the following case: I'm a Custom Task end user, I want to use `PipelineLoop`. I don't think It's possible for us to use `CustomRef` without specifying a name of an existing `PipelineLoop`.

wdyt?

Reply:
By any means, this can be a document improvement as part of https://github.com/tektoncd/pipeline/issues/5775:
- For end users, clarify when to use `unnamed CustomRun` and `named CustomRun`
- For custom task authors, clarify whether they should support `unnamed CustomRun`